src/connection/host_lookup.c

```diff
@@ -0,0 +1,670 @@
+}
+
+int
+getaddrinfo_uhost (char *node, char *service, struct addrinfo *hints, struct addrinfo **res)
+{
+  int ret = 0;
+  struct hostent *hp = NULL;
+  struct addrinfo results_out;
+  struct sockaddr_in addr_convert;
+  struct in_addr *in_addr_buf = NULL;
+
+  if (prm_get_bool_value (PRM_ID_USE_USER_HOSTS) == USE_GLIBC_HOSTS)
+    {
+      ret = getaddrinfo (node, service, hints, res);
+      goto return_phase;
+    }
+  else
+    {
+      hp = host_lookup_internal (node, NULL, HOSTNAME_TO_IPADDR);
+    }
+
+  if (hp == NULL)
+    {
+      ret = EAI_NODATA;
+      goto return_phase;
+    }
+
+  if ((in_addr_buf = (struct in_addr *) malloc (sizeof (struct in_addr))) == NULL)
+    {
+      //FREE_HOSTENT_MEM (hp);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct in_addr));
+      ret = EAI_MEMORY;
+      goto return_phase;
+    }
+
+  /*Constitute struct addrinfo for the out parameter res */
+  if (((*res) = (struct addrinfo *) malloc (sizeof (struct addrinfo))) == NULL)
+    {
+      //FREE_HOSTENT_MEM (hp);
+      free (in_addr_buf);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct addrinfo));
+      ret = EAI_MEMORY;
+      goto return_phase;
+    }
+
+  memset (&results_out, 0, sizeof (results_out));
+  if ((results_out.ai_addr = (struct sockaddr *) malloc (sizeof (struct sockaddr))) == NULL)
+    {
+      //FREE_HOSTENT_MEM (hp);
+      free (in_addr_buf);
+      freeaddrinfo (*res);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct sockaddr));
+      ret = EAI_MEMORY;
+      goto return_phase;
+    }
+
+  if (hints != NULL)
+    {
+      results_out.ai_flags = hints->ai_flags;
+      results_out.ai_family = hints->ai_family;
+      results_out.ai_socktype = hints->ai_socktype;
+      results_out.ai_protocol = IPPROTO_TCP;
+    }
+  else
+    {
+      results_out.ai_flags = (AI_V4MAPPED | AI_ADDRCONFIG);
+      results_out.ai_family = AF_UNSPEC;
+      results_out.ai_socktype = 0;
+      results_out.ai_protocol = 0;
+    }
+
+  if ((results_out.ai_canonname = (char *) malloc (sizeof (char) * HOSTNAME_LEN)) == NULL)
+    {
+      free (in_addr_buf);
+      freeaddrinfo (*res);
+      free (results_out.ai_addr);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct sockaddr));
+      ret = EAI_MEMORY;
+      goto return_phase;
+    }
+
+  memcpy (&in_addr_buf->s_addr, hp->h_addr_list[0], sizeof (in_addr_buf->s_addr));
+  memcpy (&addr_convert.sin_addr, &in_addr_buf->s_addr, sizeof (addr_convert.sin_addr));
+  memcpy (results_out.ai_addr, (struct sockaddr *) &addr_convert, sizeof (struct sockaddr));
+
+  results_out.ai_addrlen = sizeof (results_out.ai_addr);
+  results_out.ai_next = NULL;
+
+  strcpy (results_out.ai_canonname, hp->h_name);
+
+  memmove (*res, &results_out, sizeof (struct addrinfo));
+
+  FREE_MEM (in_addr_buf);
+  FREE_MEM (results_out.ai_addr);
```

Comment:
You should not You shouldn't free **results_out.ai_addr**, it will be done by caller. Please remove this line.